src/Grpc.Net.Client/Balancer/Subchannel.cs

```diff
@@ -101,25 +103,41 @@ public IDisposable OnStateChanged(Action<SubchannelState> callback)
             return registration;
         }
 
+        private bool RemoveStateChanged(StateChangedRegistration registration) => _stateChangedRegistrations.Remove(registration);
+
+        private string GetNextRegistrationId()
+        {
+            var registrationId = Interlocked.Increment(ref _currentRegistrationId);
+            return Id + "-" + registrationId;
+        }
+
         private sealed class StateChangedRegistration : IDisposable
         {
             private readonly Subchannel _subchannel;
             private readonly Action<SubchannelState> _callback;
+            private readonly string _registrationId;
 
             public StateChangedRegistration(Subchannel subchannel, Action<SubchannelState> callback)
             {
                 _subchannel = subchannel;
                 _callback = callback;
+                _registrationId = subchannel.GetNextRegistrationId();
+
+                SubchannelLog.StateChangedRegistrationCreated(_subchannel._logger, _subchannel.Id, _registrationId);
             }
 
             public void Invoke(SubchannelState state)
             {
+                SubchannelLog.ExecutingStateChangedRegistration(_subchannel._logger, _subchannel.Id, _registrationId);
```

Comment:
These are `LogLeve.Trace`, so mostly disabled (in production), but the allocation(s) from the registration-ID sill occur. If that's relevant, these allocation could be avoided by
```c#
if (_subchannel._logger.IsEnabled(LogLevel.Trace))
{
    SubchannelLog.FooBar(..., RegistrationId);
}
```
where `RegistrationId` is lazy-evaluated property that build the string from the id stored as field.

Below in the `LoggerMessage.Define` the `skipEnabledCheck` has to be set to `true`.

Reply:
That's true, but channels are long-lived and there is probably only ever one registration. It's not worth trying to save.